### PR TITLE
DAGE-23: Add llm explainability feature

### PR DESCRIPTION
**Pull request by nseidan**
Jira: https://sease.atlassian.net/browse/DAGE-23
* Added llm explanability feature  for the generated scores
* Introduced `save_llm_explanation` , `llm_explanation_destination` in configuration
* Added a unit test
* Removed e2e folder

### CODE REVIEW

File: rre-dataset-generator/src/config.py [hunk 1]
```diff
@@ -26,6 +26,8 @@ class Config(BaseModel):
     llm_configuration_file: FilePath = Field(..., description="Path to the LLM configuration file.")
     output_format: Literal['quepid', 'rre']
     output_destination: Path = Field(..., description="Path to save the output dataset.")
+    save_llm_explanation: Optional[bool] = False
+    llm_explanation_destination: Path = Field(..., description="Path to save the llm rating reasoning")
```
Comment (nicolo-rinaldi) [inline rre-dataset-generator/src/config.py]: Is this working as expected? What if the file is not there? I think the import of the config file should throw an error

File: rre-dataset-generator/dataset_generator.py
Comment (nicolo-rinaldi) [inline rre-dataset-generator/dataset_generator.py:1]: this has also been changed in the DAGE-26, depending on what PR is going to be merged first, this file needs a refactor:(

File: rre-dataset-generator/src/model/score_response.py [hunk 2]
```diff
@@ -1,14 +1,18 @@
...
-    def __init__(self, score: int, scale: str = "graded"):
+    def __init__(self, score: int, scale: str = "graded", reasoning: Optional[str] = None):
         """
         Initializes the object by validating the score.
 
         Args:
-            score: The relevance score.
-            scale: The relevance scale, either 'binary' (0-1) or 'graded' (0-2).
+            score:      The relevance score.
+            scale:      The relevance scale, either 'binary' (0-1) or 'graded' (0-2).
```
Comment (nicolo-rinaldi) [inline rre-dataset-generator/src/model/score_response.py]: I'd rather change this:
- (0-1) -> {0,1}
- (0-2) -> {0,1,2}

File: rre-dataset-generator/src/model/query_rating_context.py [hunk 3]
```diff
@@ -1,24 +1,26 @@
...
     """
 
     DOC_NOT_RATED: int = -1  # doc is not yet rated
 
     def __init__(self, query: str, doc_id: str | None = None):
         self._id: str = str(uuid.uuid4())
         self._query: str = query
-        self._doc_id_to_rating_score: Dict[str, int] = {}
-        # HANDLING NONEs - throwed error in some tests
+        self._doc_id_to_rating_score: Dict[str, Rating] = {}
```
Comment (nicolo-rinaldi) [inline rre-dataset-generator/src/model/query_rating_context.py:20]: i like this

File: rre-dataset-generator/src/config.py (hunk 1 above)
Comment (nseidan) [reply rre-dataset-generator/src/config.py]: yes, the file is created because of the llm explanation writer, but the dir `output` is already there.

File: rre-dataset-generator/src/model/score_response.py (hunk 2 above)
Comment (nseidan) [reply rre-dataset-generator/src/model/score_response.py]: Well, this is not the change I made. Not sure you or Pablo made.

File: rre-dataset-generator/src/config.py (hunk 1 above)
Comment (nicolo-rinaldi) [reply rre-dataset-generator/src/config.py]: I think maybe
```
output_path = Path(output_path)
os.makedirs(output_path.parent, exist_ok=True)
```
should be added in the method `.write()`, to avoid unexpected errors. Similar to quepid writer

File: rre-dataset-generator/src/model/score_response.py (hunk 2 above)
Comment (nicolo-rinaldi) [reply rre-dataset-generator/src/model/score_response.py]: I didn't

Comment (nseidan) [reply rre-dataset-generator/src/model/score_response.py]: since it's a documentation, i updated

File: rre-dataset-generator/src/config.py (hunk 1 above)
Comment (nseidan) [reply rre-dataset-generator/src/config.py]: updated

File: rre-dataset-generator/dataset_generator.py
Comment (nseidan) [reply rre-dataset-generator/dataset_generator.py:1]: oh i see (

**Review by nicolo-rinaldi — APPROVED** [review]

File: rre-dataset-generator/src/writers/llm_explanation_writer.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import json
+import os
+from pathlib import Path
+from typing import List, Tuple
+
+from src.writers.abstract_writer import AbstractWriter
+
+
+class LlmExplanationWriter(AbstractWriter):
```
Comment (Intrinsical-AI) [inline rre-dataset-generator/src/writers/llm_explanation_writer.py]: This is being implemented here:
https://sease.atlassian.net/browse/DAGE-18?atlOrigin=eyJpIjoiZmI5ZTliZmNjNjZmNDI4MjlmZDE2YjcxOTk1MDdjZmQiLCJwIjoiaiJ9

> Check the 
image

I don't see the need of a new Writter. Maybe adapting one of existing ones. Feel free to check with Daniele, maybe I'm mistaken, and he likes this approach

File: rre-dataset-generator/src/model/rating.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+from dataclasses import dataclass
+from typing import Optional
+
+
+@dataclass
+class Rating:
```
Comment (Intrinsical-AI) [inline rre-dataset-generator/src/model/rating.py]: why not a dict with [int,Optional[str]] ? I love dataclass, but if you planing to add a refactor would be nice

File: rre-dataset-generator/tests/unit/llm/test_llm_service_score.py [hunk 6]
```diff
@@ -29,9 +30,36 @@ def test_generate_score_with_valid_scale_EXPECTED_integer_score(scale, valid_sco
...
+    response = service.generate_score(example_doc,
+                                      query,
+                                      relevance_scale=scale,
+                                      reasoning=False)
+    assert isinstance(response, LLMScoreResponse)
+    assert response.get_score() == valid_score
+    assert response.get_reasoning() is None
+
+
+@pytest.mark.parametrize("scale, valid_score, explanation", [
```
Comment (Intrinsical-AI) [inline rre-dataset-generator/tests/unit/llm/test_llm_service_score.py:49]: Note: we're only testing "nice" functions, or "the happy path". We should add edge cases, where the app it's expected to break. That's what most of unit test should do - IMO

**Review by Intrinsical-AI — COMMENTED** [review]
The test updates and cleanup look good. However, the LLM reasoning implementation seems too complex for what it needs. Open to discussing the reasoning behind this design.

### Data persistence

This [PR](https://github.com/SeaseLtd/rated-ranking-evaluator/blob/02fa354d7ae91bd7e78efa37de0d5d408dad09ff/rre-dataset-generator/src/search_engine/data_store.py) should be approved soon. It includes new persistence functions in `data_store`, which align 100% with your task but go a bit further.

* I don’t think a new writer is needed. The reasoning dump could be handled by slightly extending `data_store` and `LLM_service` (e.g., adding a `reasoning` param). See linked PR for context.

### Use of dataclasses

Not sure about the choice to use a `dataclass` for `Rating`, especially after pushing to remove the Pydantic version. It’s a simple `dict[int, Optional[str]]`.

* Why use a dataclass here?
* What advantages does it bring over Pydantic?
* Why not apply the same logic to more complex models like `Query_Rating_Context`, as I proposed since the beginning, which would benefit more from structured classes?

I love dataclasses, but mixing patterns without strong reasoning, adds inconsistency. Open to hearing you.

**Comment by Intrinsical-AI** [comment]
Note: I keep the previous feedback, but just realized maybe you were not aware on the [PR](https://github.com/SeaseLtd/rated-ranking-evaluator/blob/02fa354d7ae91bd7e78efa37de0d5d408dad09ff/rre-dataset-generator/src/search_engine/data_store.py), which now implements saving / loading functions for a common json file, since it's recent (past Friday)

File: rre-dataset-generator/src/model/rating.py (hunk 5 above)
Comment (dantuzi) [reply rre-dataset-generator/src/model/rating.py]: why are you proposing a Dict? we have a single rating value

Comment (Intrinsical-AI) [reply rre-dataset-generator/src/model/rating.py]: You previously mentioned that using a full Pydantic model might be overkill for Rating. If we agree on that, then a plain dictionary seems like a lightweight and flexible possible alternative. For example:

```
rating: dict = {
    "rating_id": "xyz123",
    "reasoning": "The document is relevant to the query because it describes details of the incident in Bogotá",
    "score": 2
}
```

or for multi:
```
ratings: dict = {
    "xyz123": {
        "reasoning": "The document is relevant to the query because it describes details of the incident in Bogotá",
        "score": 2
    },
    "abc456": {
        "reasoning": "It partially matches the query topic",
        "score": 1
    }
}
```

So I approve the current approach, I'm just discussing alternatives based on previous feedback

Comment (nseidan) [reply rre-dataset-generator/src/model/rating.py]: We don't have id for rating.  Encapsulation for fields into a class is compliant with OOP.

File: rre-dataset-generator/src/writers/llm_explanation_writer.py (hunk 4 above)
Comment (nseidan) [reply rre-dataset-generator/src/writers/llm_explanation_writer.py]: correct, I just aligned with the persistent data storage.

File: rre-dataset-generator/src/model/rating.py (hunk 5 above)
Comment (nseidan) [reply rre-dataset-generator/src/model/rating.py]: I updated Rating object to use Pydantic model to keep consistency along with Document/QueryContextRating models.

File: rre-dataset-generator/tests/unit/llm/test_llm_service_score.py (hunk 6 above)
Comment (nseidan) [reply rre-dataset-generator/tests/unit/llm/test_llm_service_score.py:49]: added

Comment (Intrinsical-AI) [reply rre-dataset-generator/tests/unit/llm/test_llm_service_score.py:49]: Good! :+1: 

Thanks

File: rre-dataset-generator/src/writers/llm_explanation_writer.py (hunk 4 above)
Comment (Intrinsical-AI) [reply rre-dataset-generator/src/writers/llm_explanation_writer.py]: :+1:

File: rre-dataset-generator/src/model/rating.py (hunk 5 above)
Comment (Intrinsical-AI) [reply rre-dataset-generator/src/model/rating.py]: I agree :100:, good aligment

**Review by Intrinsical-AI — APPROVED** [review]

File: rre-dataset-generator/dataset_generator.py [hunk 7]
```diff
@@ -54,28 +55,31 @@ def generate_and_add_queries(llm_service: LLMService, config: Config, data_store
             if len(data_store.get_queries()) >= config.num_queries_needed:
                 return
             query_id: str = data_store.add_query(query_text, doc.id)
-            data_store.add_rating_score(query_id, doc.id, max(config.relevance_label_set))
+            data_store.add_rating_score(query_id, doc.id, max(config.relevance_label_set),
+                                        "Human Reasoning: max rating score is assigned when generated query")
```
Comment (dantuzi) [inline rre-dataset-generator/dataset_generator.py]: Default max rating is assigned because the query is generated by the document

File: rre-dataset-generator/src/llm/llm_service.py [hunk 8]
```diff
@@ -84,15 +96,16 @@ def generate_score(self, document: Document, query: str, relevance_scale: str) -
 
         try:
             score = json.loads(raw)['score']
+            score_reasoning = None
+            if reasoning:
```
Comment (dantuzi) [inline rre-dataset-generator/src/llm/llm_service.py]: I would rename readinging to exiplaination or something like that

File: rre-dataset-generator/tests/unit/llm/test_llm_service_score.py [hunk 9]
```diff
@@ -29,9 +30,43 @@ def test_generate_score_with_valid_scale_EXPECTED_integer_score(scale, valid_sco
...
+    with pytest.raises(ValueError, match="Invalid LLM response:"):
+        service.generate_score(example_doc, "query", relevance_scale="binary", reasoning=True)
+
+
+@pytest.mark.parametrize("scale, valid_score, explanation", [
+    ("graded", 0, "The query is clearly not about cars."),
+    ("graded", 1, "Camry is a car, so it is relevant."),
+    ("graded", 2, "This exactly matches the definition of 'car of the year'."),
+])
+def test_generate_score_with_reasoning_returns_explanation(scale, valid_score, explanation, example_doc):
```
Comment (dantuzi) [inline rre-dataset-generator/tests/unit/llm/test_llm_service_score.py]: `<what you test>__expect__<what you expect>`

File: rre-dataset-generator/tests/unit/llm/test_llm_service_score.py [hunk 10]
```diff
@@ -45,7 +80,8 @@ def test_generate_score_with_valid_scale_EXPECTED_integer_score(scale, valid_sco
     ('graded', '{"score": 1.5}', 'Score must be 0, 1, or 2 for graded scale, got 1.5'),
     ('graded', '{"score": null}', 'Score must be 0, 1, or 2 for graded scale, got None'),
 ])
-def test_generate_score_with_invalid_llm_responses_EXPECTED_value_error(scale, response_json, expected_error, example_doc):
+def test_generate_score_with_invalid_llm_responses_EXPECTED_value_error(scale, response_json, expected_error,
```
Comment (dantuzi) [inline rre-dataset-generator/tests/unit/llm/test_llm_service_score.py]: `__expect__` lowercase

File: rre-dataset-generator/src/search_engine/data_store.py [hunk 11]
```diff
@@ -202,3 +203,25 @@ def load_tmp_file_content(self) -> None:
         documents = file_content.get("documents", {})
         for doc_id, doc_data in documents.items():
             self.add_document(doc_id, Document.model_validate(doc_data))
+
+    def export_all_records_with_reasoning(self, output_path: str | Path) -> None:
```
Comment (dantuzi) [inline rre-dataset-generator/src/search_engine/data_store.py]: Can we have a test for this functionality?

**Review by dantuzi — CHANGES_REQUESTED** [review]

File: rre-dataset-generator/dataset_generator.py (hunk 7 above)
Comment (nseidan) [reply rre-dataset-generator/dataset_generator.py]: updated

File: rre-dataset-generator/src/llm/llm_service.py (hunk 8 above)
Comment (nseidan) [reply rre-dataset-generator/src/llm/llm_service.py]: changed to `explanation`

File: rre-dataset-generator/tests/unit/llm/test_llm_service_score.py (hunk 9 above)
Comment (nseidan) [reply rre-dataset-generator/tests/unit/llm/test_llm_service_score.py]: updated

File: rre-dataset-generator/tests/unit/llm/test_llm_service_score.py (hunk 10 above)
Comment (nseidan) [reply rre-dataset-generator/tests/unit/llm/test_llm_service_score.py]: updated

File: rre-dataset-generator/src/search_engine/data_store.py (hunk 11 above)
Comment (nseidan) [reply rre-dataset-generator/src/search_engine/data_store.py]: added

**Review by dantuzi — APPROVED** [review]
